allennlp/models/coreference_resolution/coref.py

```diff
@@ -0,0 +1,824 @@
+        top_span_embeddings = util.batched_index_select(span_embeddings,
+                                                        top_span_indices,
+                                                        flat_top_span_indices)
+        # Shape: (batch_size, num_spans_to_keep, 1)
+        # TODO(Mark): If we parameterised the mention scorer to score things in (0, inf)
+        # I think we could get rid of the need for this mask entirely.
+        top_span_mask = util.batched_index_select(span_mask,
+                                                  top_span_indices,
+                                                  flat_top_span_indices)
+        top_span_mention_scores = util.batched_index_select(mention_scores,
+                                                            top_span_indices,
+                                                            flat_top_span_indices)
+        top_span_starts = util.batched_index_select(span_starts,
+                                                    top_span_indices,
+                                                    flat_top_span_indices)
+        top_span_ends = util.batched_index_select(span_ends,
+                                                  top_span_indices,
+                                                  flat_top_span_indices)
+
+        # Compute indices for antecedent spans to consider.
+        max_antecedents = min(self._max_antecedents, num_spans_to_keep)
+
+        # This next section can be confusing - the returned tensors do _not_ have a batch size
+        # dimension. This is because the antecedent indices are agnostic to the batch size -
+        # it is a function only of the number of spans we are considering (i.e which survived the
+        # pruning stage) and the max_antecedents that we will consider. We then _look up_
+        # these indices in top_span_embeddings to get the embeddings for them for each
+        # instance in the batch. This tensor represents the indices which we will be generating
+        # distributions over - for each span we kept, we will generate a distribution over
+        # the maximum number of antecedents we choose to consider. As these are just the
+        # indices of the spans, and we consider the same number for each element in the batch,
+        # it is batch agnostic. In fact, the next step in the model is to look up all of the
+        # embeddings for these indices in _every_ batch.
+
+        # Shapes:
+        # (num_spans_to_keep, max_antecedents),
+        # (1, max_antecedents),
+        # (1, num_spans_to_keep, max_antecedents)
+        antecedent_indices, antecedent_offsets, antecedent_log_mask = \
+            self._generate_antecedents(num_spans_to_keep, max_antecedents, text_mask.is_cuda)
+        # Select tensors relating to the antecedent spans.
+        # Shape: (batch_size, num_spans_to_keep, max_antecedents, embedding_size)
+        candidate_antecedent_embedidngs = util.flattened_index_select(top_span_embeddings,
```

Comment:
`flattened` doesn't seem like the right name here.  Maybe call this `util.2d_index_select` instead?  You could even have the dimension be an argument to the function (it's currently hard-coded to 1).